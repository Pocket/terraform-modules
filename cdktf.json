{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
