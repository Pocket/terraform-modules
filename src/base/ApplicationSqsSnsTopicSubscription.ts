import { Resource, TerraformMetaArguments, TerraformResource } from 'cdktf';
import { Construct } from 'constructs';
import { sqs, sns, iam } from '@cdktf/provider-aws';
import { SnsTopicSubscriptionConfig } from '@cdktf/provider-aws/lib/sns';

export interface ApplicationSqsSnsTopicSubscriptionProps
  extends TerraformMetaArguments {
  name: string;
  snsTopicArn: string;
  sqsQueue: sqs.SqsQueue | sqs.DataAwsSqsQueue;
  snsDlq?: sqs.SqsQueue;
  tags?: { [key: string]: string };
  dependsOn?: TerraformResource[];
}

/**
 * Creates an SNS to SQS subscription
 */
export class ApplicationSqsSnsTopicSubscription extends Resource {
  public readonly snsTopicSubscription: sns.SnsTopicSubscription;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationSqsSnsTopicSubscriptionProps
  ) {
    super(scope, name);

    const snsTopicDlq = this.config.snsDlq ?? this.createSqsSubscriptionDlq();
    this.snsTopicSubscription = this.createSnsTopicSubscription(snsTopicDlq);
    this.createPoliciesForSnsToSQS(snsTopicDlq);
  }

  /**
   * Create a dead-letter queue for failed SNS messages
   * @private
   */
  private createSqsSubscriptionDlq(): sqs.SqsQueue {
    return new sqs.SqsQueue(this, 'sns-topic-dql', {
      name: `${this.config.name}-SNS-Topic-DLQ`,
      tags: this.config.tags,
      provider: this.config.provider,
    });
  }

  /**
   * Create an SNS subscription for SQS
   * @param snsTopicDlq
   * @private
   */
  private createSnsTopicSubscription(
    snsTopicDlq: sqs.SqsQueue
  ): sns.SnsTopicSubscription {
    return new sns.SnsTopicSubscription(this, 'sns-subscription', {
      topicArn: this.config.snsTopicArn,
      protocol: 'sqs',
      endpoint: this.config.sqsQueue.arn,
      redrivePolicy: JSON.stringify({
        deadLetterTargetArn: snsTopicDlq.arn,
      }),
      dependsOn: [
        snsTopicDlq,
        ...(this.config.dependsOn ? this.config.dependsOn : []),
      ],
      provider: this.config.provider,
    } as SnsTopicSubscriptionConfig);
  }

  /**
   * Create IAM policies to allow SNS to write to the target SQS queue and a
   * dead-letter queue
   * @param snsTopicDlq
   * @private
   */
  private createPoliciesForSnsToSQS(snsTopicDlq: sqs.SqsQueue): void {
    [
      { name: 'sns-sqs', resource: this.config.sqsQueue },
      { name: 'sns-dlq', resource: snsTopicDlq },
    ].forEach((queue) => {
      const policy = new iam.DataAwsIamPolicyDocument(
        this,
        `${queue.name}-policy-document`,
        {
          statement: [
            {
              effect: 'Allow',
              actions: ['sqs:SendMessage'],
              resources: [queue.resource.arn],
              principals: [
                {
                  identifiers: ['sns.amazonaws.com'],
                  type: 'Service',
                },
              ],
              condition: [
                {
                  test: 'ArnEquals',
                  variable: 'aws:SourceArn',
                  values: [this.config.snsTopicArn],
                },
              ],
            },
          ],
          dependsOn: [queue.resource] as TerraformResource[],
          provider: this.config.provider,
        }
      ).json;

      new sqs.SqsQueuePolicy(this, `${queue.name}-policy`, {
        queueUrl: queue.resource.url,
        policy: policy,
        provider: this.config.provider,
      });
    });
  }
}
