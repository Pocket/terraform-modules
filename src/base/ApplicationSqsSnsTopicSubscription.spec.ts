import { SqsQueue } from '@cdktf/provider-aws/lib/sqs-queue';
import { Testing } from 'cdktf';
import { ApplicationSqsSnsTopicSubscription } from './ApplicationSqsSnsTopicSubscription';

describe('ApplicationSqsSnsTopicSubscription', () => {
  const getConfig = (stack) => ({
    name: 'test-sns-subscription',
    snsTopicArn: 'arn:aws:sns:TopicName',
    sqsQueue: new SqsQueue(stack, 'sqs', {
      name: 'test-sqs',
    }),
  });

  const getConfigWithDlq = (stack) => ({
    name: 'test-sns-subscription',
    snsTopicArn: 'arn:aws:sns:TopicName',
    sqsQueue: new SqsQueue(stack, 'sqs', {
      name: 'test-sqs',
    }),
    snsDlq: new SqsQueue(stack, 'dlq', {
      name: 'test-sqs-dlq',
    }),
  });

  it('renders an SQS SNS subscription without tags', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationSqsSnsTopicSubscription(
        stack,
        'sqs-sns-subscription',
        getConfig(stack),
      );
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders an SQS SNS subscription with tags', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationSqsSnsTopicSubscription(stack, 'sqs-sns-subscription', {
        ...getConfig(stack),
        tags: { hello: 'there' },
      });
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders an SQS SNS subscription witg dlq passed', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationSqsSnsTopicSubscription(stack, 'sqs-sns-subscription', {
        ...getConfigWithDlq(stack),
        tags: { hello: 'there' },
      });
    });
    expect(synthed).toMatchSnapshot();
  });
});
