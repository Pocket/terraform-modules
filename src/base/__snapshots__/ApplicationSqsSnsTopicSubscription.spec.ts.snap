// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationSqsSnsTopicSubscription renders an SQS SNS subscription witg dlq passed 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "sqs-sns-subscription_sns-dlq-policy-document_46022E28": {
        "depends_on": [
          "aws_sqs_queue.dlq"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.dlq.arn}"
            ]
          }
        ]
      },
      "sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA": {
        "depends_on": [
          "aws_sqs_queue.sqs"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.sqs.arn}"
            ]
          }
        ]
      }
    }
  },
  "resource": {
    "aws_sns_topic_subscription": {
      "sqs-sns-subscription_5757EDDD": {
        "depends_on": [
          "aws_sqs_queue.dlq"
        ],
        "endpoint": "\${aws_sqs_queue.sqs.arn}",
        "protocol": "sqs",
        "redrive_policy": "{\\"deadLetterTargetArn\\":\\"\${aws_sqs_queue.dlq.arn}\\"}",
        "topic_arn": "arn:aws:sns:TopicName"
      }
    },
    "aws_sqs_queue": {
      "dlq": {
        "name": "test-sqs-dlq"
      },
      "sqs": {
        "name": "test-sqs"
      }
    },
    "aws_sqs_queue_policy": {
      "sqs-sns-subscription_sns-dlq-policy_736EBF83": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-dlq-policy-document_46022E28.json}",
        "queue_url": "\${aws_sqs_queue.dlq.url}"
      },
      "sqs-sns-subscription_sns-sqs-policy_ADC74422": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA.json}",
        "queue_url": "\${aws_sqs_queue.sqs.url}"
      }
    }
  }
}"
`;

exports[`ApplicationSqsSnsTopicSubscription renders an SQS SNS subscription with tags 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "sqs-sns-subscription_sns-dlq-policy-document_46022E28": {
        "depends_on": [
          "aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.arn}"
            ]
          }
        ]
      },
      "sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA": {
        "depends_on": [
          "aws_sqs_queue.sqs"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.sqs.arn}"
            ]
          }
        ]
      }
    }
  },
  "resource": {
    "aws_sns_topic_subscription": {
      "sqs-sns-subscription_5757EDDD": {
        "depends_on": [
          "aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678"
        ],
        "endpoint": "\${aws_sqs_queue.sqs.arn}",
        "protocol": "sqs",
        "redrive_policy": "{\\"deadLetterTargetArn\\":\\"\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.arn}\\"}",
        "topic_arn": "arn:aws:sns:TopicName"
      }
    },
    "aws_sqs_queue": {
      "sqs": {
        "name": "test-sqs"
      },
      "sqs-sns-subscription_sns-topic-dql_23368678": {
        "name": "test-sns-subscription-SNS-Topic-DLQ",
        "tags": {
          "hello": "there"
        }
      }
    },
    "aws_sqs_queue_policy": {
      "sqs-sns-subscription_sns-dlq-policy_736EBF83": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-dlq-policy-document_46022E28.json}",
        "queue_url": "\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.url}"
      },
      "sqs-sns-subscription_sns-sqs-policy_ADC74422": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA.json}",
        "queue_url": "\${aws_sqs_queue.sqs.url}"
      }
    }
  }
}"
`;

exports[`ApplicationSqsSnsTopicSubscription renders an SQS SNS subscription without tags 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "sqs-sns-subscription_sns-dlq-policy-document_46022E28": {
        "depends_on": [
          "aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.arn}"
            ]
          }
        ]
      },
      "sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA": {
        "depends_on": [
          "aws_sqs_queue.sqs"
        ],
        "statement": [
          {
            "actions": [
              "sqs:SendMessage"
            ],
            "condition": [
              {
                "test": "ArnEquals",
                "values": [
                  "arn:aws:sns:TopicName"
                ],
                "variable": "aws:SourceArn"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "sns.amazonaws.com"
                ],
                "type": "Service"
              }
            ],
            "resources": [
              "\${aws_sqs_queue.sqs.arn}"
            ]
          }
        ]
      }
    }
  },
  "resource": {
    "aws_sns_topic_subscription": {
      "sqs-sns-subscription_5757EDDD": {
        "depends_on": [
          "aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678"
        ],
        "endpoint": "\${aws_sqs_queue.sqs.arn}",
        "protocol": "sqs",
        "redrive_policy": "{\\"deadLetterTargetArn\\":\\"\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.arn}\\"}",
        "topic_arn": "arn:aws:sns:TopicName"
      }
    },
    "aws_sqs_queue": {
      "sqs": {
        "name": "test-sqs"
      },
      "sqs-sns-subscription_sns-topic-dql_23368678": {
        "name": "test-sns-subscription-SNS-Topic-DLQ"
      }
    },
    "aws_sqs_queue_policy": {
      "sqs-sns-subscription_sns-dlq-policy_736EBF83": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-dlq-policy-document_46022E28.json}",
        "queue_url": "\${aws_sqs_queue.sqs-sns-subscription_sns-topic-dql_23368678.url}"
      },
      "sqs-sns-subscription_sns-sqs-policy_ADC74422": {
        "policy": "\${data.aws_iam_policy_document.sqs-sns-subscription_sns-sqs-policy-document_ABFC60AA.json}",
        "queue_url": "\${aws_sqs_queue.sqs.url}"
      }
    }
  }
}"
`;
